File: minigit-4/ts/src/commands/ls-tree.ts
// ls-tree command - List tree contents

import { findRepoRoot } from '../utils';
import { readObject, parseCommitContent, parseTreeContent, objectExists } from '../objects';
import { resolveRevision } from '../refs';

export function lsTree(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let recursive = false;
  let nameOnly = false;
  let treeIsh: string | null = null;

  for (const arg of args) {
    if (arg === '-r') {
      recursive = true;
    } else if (arg === '--name-only') {
      nameOnly = true;
    } else if (!arg.startsWith('-')) {
      treeIsh = arg;
    }
  }

  if (!treeIsh) {
    console.error('fatal: tree-ish required');
    return 1;
  }

  // Resolve to tree SHA
  let treeSha = resolveRevision(repoRoot, treeIsh);
  if (!treeSha || !objectExists(repoRoot, treeSha)) {
    console.error(`fatal: not a valid object name ${treeIsh}`);
    return 1;
  }

  // If it's a commit, get its tree
  const { type, content } = readObject(repoRoot, treeSha);
  if (type === 'commit') {
    const commitInfo = parseCommitContent(content);
    treeSha = commitInfo.tree;
  } else if (type !== 'tree') {
    console.error(`fatal: not a tree object`);
    return 1;
  }

  listTree(repoRoot, treeSha, '', recursive, nameOnly);
  return 0;
}

function listTree(repoRoot: string, treeSha: string, prefix: string, recursive: boolean, nameOnly: boolean): void {
  const { content } = readObject(repoRoot, treeSha);
  const entries = parseTreeContent(content);

  for (const entry of entries) {
    const fullName = prefix ? `${prefix}/${entry.name}` : entry.name;
    const isTree = entry.mode === '40000';
    const typeStr = isTree ? 'tree' : 'blob';

    if (recursive && isTree) {
      // Don't print tree entry itself when recursive, just recurse into it
      listTree(repoRoot, entry.sha, fullName, recursive, nameOnly);
    } else if (nameOnly) {
      console.log(fullName);
    } else {
      const modeStr = entry.mode.padStart(6, '0');
      console.log(`${modeStr} ${typeStr} ${entry.sha}\t${fullName}`);
    }
  }
}

File: minigit-4/py/commands/ls_tree.py
"""ls-tree command - List tree contents."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import read_object, Tree, resolve_object


def run(args):
    """List tree contents."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    recursive = False
    name_only = False
    target = None

    for arg in args:
        if arg == '-r':
            recursive = True
        elif arg == '--name-only':
            name_only = True
        elif not arg.startswith('-'):
            target = arg

    if not target:
        print("fatal: no tree specified", file=sys.stderr)
        return 1

    # Resolve to tree
    try:
        sha = resolve_object(target, repo_root)
    except ValueError as e:
        print(f"fatal: Not a valid object name {target}", file=sys.stderr)
        return 1

    # Get tree SHA
    obj_type, data = read_object(sha, repo_root)

    if obj_type == 'commit':
        from objects import Commit
        commit = Commit.deserialize(data)
        tree_sha = commit.tree_sha
    elif obj_type == 'tree':
        tree_sha = sha
    else:
        print(f"fatal: not a tree object", file=sys.stderr)
        return 1

    # List tree
    list_tree(tree_sha, '', repo_root, recursive, name_only)
    return 0


def list_tree(tree_sha: str, prefix: str, repo_root: Path,
              recursive: bool, name_only: bool):
    """Recursively list tree contents."""
    _, data = read_object(tree_sha, repo_root)
    tree = Tree.deserialize(data)

    for entry in tree.entries:
        path = f'{prefix}{entry.name}' if prefix else entry.name

        if entry.mode.startswith('40'):
            entry_type = 'tree'
        else:
            entry_type = 'blob'

        if name_only:
            if entry_type == 'tree' and recursive:
                # Don't print directory itself in recursive name-only mode
                pass
            else:
                print(path)
        else:
            print(f'{entry.mode} {entry_type} {entry.sha}\t{path}')

        if recursive and entry_type == 'tree':
            list_tree(entry.sha, path + '/', repo_root, recursive, name_only)
